src/poseidon/mod.rs

```diff
@@ -0,0 +1,377 @@
+        let alpha = params.alpha;
+
+        let mds = params.mds.clone();
+
+        let ark = params.ark.to_vec();
+
+        let rate = 2;
+        let capacity = 1;
+        let state = vec![F::zero(); rate + capacity];
+        let mode = PoseidonSpongeMode::Absorbing {
+            next_absorb_index: 0,
+        };
+
+        Self {
+            full_rounds,
+            partial_rounds,
+            alpha,
+            ark,
+            mds,
+
+            state,
+            rate,
+            capacity,
+            mode,
+        }
+    }
+
+    fn absorb(&mut self, input: &impl Absorb) {
+        let elems = input.to_sponge_field_elements_as_vec::<F>();
+        if elems.is_empty() {
+            return;
+        }
+
+        match self.mode {
+            PoseidonSpongeMode::Absorbing { next_absorb_index } => {
+                let mut absorb_index = next_absorb_index;
+                if absorb_index == self.rate {
+                    self.permute();
+                    absorb_index = 0;
+                }
+                self.absorb_internal(absorb_index, elems.as_slice());
+            }
+            PoseidonSpongeMode::Squeezing {
+                next_squeeze_index: _,
+            } => {
+                self.permute();
+                self.absorb_internal(0, elems.as_slice());
+            }
+        };
+    }
+
+    fn squeeze_bytes(&mut self, num_bytes: usize) -> Vec<u8> {
+        let usable_bytes = (F::Params::CAPACITY / 8) as usize;
+
+        let num_elements = (num_bytes + usable_bytes - 1) / usable_bytes;
+        let src_elements = self.squeeze_native_field_elements(num_elements);
+
+        let mut bytes: Vec<u8> = Vec::with_capacity(usable_bytes * num_elements);
+        for elem in &src_elements {
+            let elem_bytes = elem.into_repr().to_bytes_le();
+            bytes.extend_from_slice(&elem_bytes[..usable_bytes]);
+        }
+
+        bytes.truncate(num_bytes);
+        bytes
+    }
+
+    fn squeeze_bits(&mut self, num_bits: usize) -> Vec<bool> {
+        let usable_bits = F::Params::CAPACITY as usize;
+
+        let num_elements = (num_bits + usable_bits - 1) / usable_bits;
+        let src_elements = self.squeeze_native_field_elements(num_elements);
+
+        let mut bits: Vec<bool> = Vec::with_capacity(usable_bits * num_elements);
+        for elem in &src_elements {
+            let elem_bits = elem.into_repr().to_bits_le();
+            bits.extend_from_slice(&elem_bits[..usable_bits]);
+        }
+
+        bits.truncate(num_bits);
+        bits
+    }
+
+    fn squeeze_field_elements<F2: PrimeField>(&mut self, num_elements: usize) -> Vec<F2> {
+        if TypeId::of::<F>() == TypeId::of::<F2>() {
+            let result = self.squeeze_native_field_elements(num_elements);
+            let mut casted = Vec::with_capacity(result.len());
+            batch_field_cast(&result, &mut casted);
+            casted
+        } else {
+            self.squeeze_field_elements_with_sizes::<F2>(
```

Comment:
shouldn't we also implement the same logic in `squeeze_field_elements_with_sizes`? (Sorry if that's already been done, I wasn't able to find it.)

Reply:
I think so (same as https://github.com/arkworks-rs/sponge/pull/13#discussion_r631349338)

Reply:
fixed